ALC5682 machine driver upgrade

brentlu opened the pull request:
Some changes to ALC5682 machine driver:

6f6e5fb00d8c ASoC: Intel: sof-rt5682: add mclk_en to sof_rt5682_private
4a952a69aa5e ASoC: Intel: sof-rt5682: remove SOF_RT5682_MCLK_BYTCHT_EN
=> code refactoring, no functional change

78fa9b265b5d ASoC: Intel: sof_rt5682: add icl_rt5682_def for icl boards
=> add board id for icl platform

3b7e2eaaffe4 ASoC: Intel: sof-rt5682: add driver_data to sof_rt5682 board
=> code refactoring and also fixes a bug of Minnowboard using bclk as clock source

8df5e9c29f59 ASoC: Intel: sof-rt5682: setup pll_id only when needed
=> code refactoring, no functional change

5737a1f18dcb ASoC: Intel: sof-rt5682: get bclk frequency from topology
5b4f666e53cc ASoC: Intel: sof-rt5682: support bclk as PLL source on rt5682s
=> support using 3.072MHz bclk as clock source

Tested on mtl chromebook.

brentlu committed:
ASoC: Intel: sof-rt5682: add mclk_en to sof_rt5682_private

Add a variable mclk_en to sof_rt5682_private structure to reduce
global variable access. No functional change.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: sof-rt5682: remove SOF_RT5682_MCLK_BYTCHT_EN

We don't need this quirk flag since 'is_legacy_cpu' will be true if
this is a BYT/CHT board.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: sof-rt5682: add driver_data to sof_rt5682 board

Move default BYT/CHT quirk to driver_data of sof_rt5682 board. This
fixes a problem that DMI quirk of Minnowboard board got overwritten in
probe function since it's a BYT board.

Fixes: c68e07970eca ("ASoC: intel: sof_rt5682: Add quirk for number of HDMI DAI's")
Signed-off-by: Brent Lu <brent.lu@intel.com>

bardliao commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -801,6 +792,8 @@ static int sof_audio_probe(struct platform_device *pdev)
 	{
 		.name = "sof_rt5682",
+		.driver_data = (kernel_ulong_t)(SOF_RT5682_MCLK_EN |
+					SOF_SSP_PORT_CODEC(2)),
```
SOF_RT5682_MCLK_EN | SOF_SSP_PORT_CODEC(2) is only for BYT and CHT, but "sof_rt5682" is also for ICL. Won't it impact ICL?

brentlu replied:
The DMI quirk should assign the driver_data for ICL. Or we could add a new board id like 'icl_rt5682_def' and remove this quirk?

```
	{
		.callback = sof_rt5682_quirk_cb,
		.matches = {
			DMI_MATCH(DMI_SYS_VENDOR, "Intel Corporation"),
			DMI_MATCH(DMI_PRODUCT_NAME, "Ice Lake Client"),
		},
		.driver_data = (void *)(SOF_RT5682_MCLK_EN |
					SOF_SSP_PORT_CODEC(0)),
	},
```

bardliao replied:
Right, but the quirk is only for ICL RVP. What about other ICL products?

brentlu replied:
```
static unsigned long sof_rt5682_quirk = SOF_RT5682_MCLK_EN |
					SOF_SSP_PORT_CODEC(0);
```

They will use default value in original code. After checking dashboard, ICL program was cancelled. Maybe we could:

1. Remove DMI quirk of ICL RVP directly.
2. Or add a icl_rt5682_def board which is using SSP0 for codec and remove the DMI quirk.

Any idea?

bardliao replied:
> ```
> static unsigned long sof_rt5682_quirk = SOF_RT5682_MCLK_EN |
> 					SOF_SSP_PORT_CODEC(0);
> ```
> 
> They will use default value in original code. After checking dashboard, ICL program was cancelled. Maybe we could:

The default value is
```
static unsigned long sof_rt5682_quirk = SOF_RT5682_MCLK_EN |
                                        SOF_SSP_PORT_CODEC(0);
```
But, after adding ".driver_data = (kernel_ulong_t)(SOF_RT5682_MCLK_EN | SOF_SSP_PORT_CODEC(2))," to the "sof_rt5682" id, sof_rt5682_quirk will be overwritten by sof_audio_probe() for the machines that use "sof_rt5682" machine driver which includes ICL platforms with rt5682 codec.
```
        if (pdev->id_entry && pdev->id_entry->driver_data)
                sof_rt5682_quirk = (unsigned long)pdev->id_entry->driver_data;
```

> 
> 1. Remove DMI quirk of ICL RVP directly.
> 2. Or add a icl_rt5682_def board which is using SSP0 for codec and remove the DMI quirk.
> 

I would choose option 2. And do the corresponding change in snd_soc_acpi_intel_icl_machines[].

> Any idea?

brentlu replied:
done

bardliao commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -364,18 +364,28 @@ static int sof_rt5682_hw_params(struct snd_pcm_substream *substream,
+			case 2400000:
+			default:
+				dev_err(rtd->dev, "invalid pll_in freq %d\n", pll_in);
+				return -EINVAL;
```
We don't need to return error here. Let codec driver to determine if it can support the pll_in or not. So that if someday codec driver adds support, we don't need to edit machine driver.
I would suggest we list the pll_in that need to use PLL1 and others use PLL2 by default.

brentlu replied:
done

brentlu committed:
ASoC: Intel: sof-rt5682: get bclk frequency from topology

A different bclk frequency 3.072MHz was introduced to tgl platform and
is used in mtl topologies. Use SOF API to get frequency from topology
instead of hardcoding.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: sof-rt5682: setup pll_id only when needed

The variable 'pll_id' is needed only when we use snd_soc_dai_set_pll()
to setup PLL. Move the code segment to improve some readability.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: sof-rt5682: support bclk as PLL source on rt5682s

For rt5682s codec, we could use bclk as PLL source when the frequency
is 3.072MHz but no 2.4MHz. Update the code to select correct pll_id
and clk_id for 3.072MHz bclk.

Signed-off-by: Brent Lu <brent.lu@intel.com>

plbossart commented:
let's wait for the GitHub compile results to show up, they failed initially for some reason.

brentlu committed:
ASoC: Intel: sof_rt5682: add icl_rt5682_def for icl boards

Add the board config icl_rt5682_def to rt5682 machine driver for all
icl boards using default SSP port allocation (headphone codec on SSP0).

Signed-off-by: Brent Lu <brent.lu@intel.com>
